### ISSUE TITLE
Support `tls-server-name` field of the kubeconfig cluster configuration

### ISSUE THREAD

**Issue by AntonAM** [kube-rs/kube-rs#991]
### Would you like to work on this feature?

maybe

### What problem are you trying to solve?

Currently `kube-rs` doesn't support `tls-server-name` in the kubeconfig and just ignores it. That doesn't allow to specify correct host in the SNI. We felt it when using `kubectl-view-allocation` k8s plugin with Teleport (issue there: https://github.com/gravitational/teleport/issues/15106 ), where correct server name in the SNI is required to reach right endpoint, since multiple services are multiplexed on one port.

### Describe the solution you'd like

Support `tls-server-name` field as per k8s docs - it should be transferred as part of SNI during TSL handshake.

### Describe alternatives you've considered

-

### Documentation, Adoption, Migration Strategy

https://kubernetes.io/docs/reference/config-api/client-authentication.v1/

### Target crate for feature

kube-client

**Comment by clux** [comment]
This is (afaikt) a feature for authentication for [`RefreshableToken::Exec`](https://github.com/kube-rs/kube-rs/blob/251728e49d4ebf0416c0d9d84361176169990dc7/kube-client/src/client/auth/mod.rs#L140-L156) bubbling down through our `Config` from the Kubeconfig's [`Cluster`](https://github.com/kube-rs/kube-rs/blob/251728e49d4ebf0416c0d9d84361176169990dc7/kube-client/src/config/file_config.rs#L84-L109) (where it is [supposed to be](https://github.com/kubernetes/client-go/blob/46d4284b93f0f2fdf846e2593d05c8094689fabc/tools/clientcmd/api/v1/types.go#L62-L70).

If I understand it right, it involves extending our [`ExecCredential`](https://github.com/kube-rs/kube-rs/blob/251728e49d4ebf0416c0d9d84361176169990dc7/kube-client/src/client/auth/mod.rs#L435-L444) (in particular the spec), and then do something with it in the [`auth_exec`](https://github.com/kube-rs/kube-rs/blob/251728e49d4ebf0416c0d9d84361176169990dc7/kube-client/src/client/auth/mod.rs#L463-L488) fn? I'm not really sure what.

I can see some [light uses of it in client-go](https://github.com/kubernetes/client-go/search?p=1&q=TLSServerName) but it doesn't really explain. Help would be appreciated with this one.

**Comment by MikailBag** [comment]
As I see, client-go passes tls-server-name to the underlying TLS layer here:
https://github.com/kubernetes/client-go/blob/037b5fd01db4432f8f6bb6139ff2b6dd00008b99/transport/transport.go#L94

Documentation for `crypto/tls.Config` is https://pkg.go.dev/crypto/tls#Config

This specific field is documented as follows:

> ServerName is used to verify the hostname on the returned certificates unless InsecureSkipVerify is given. It is also included in the client's handshake to support virtual hosting unless it is an IP address.

**Comment by AntonAM** [comment]
@MikailBag yep, you are right, it's and server name to provide to TLS layer, when it connects to the server. And so it can perform necessary checks and SNI can be used correctly 👍

**Comment by clux** [comment]
This is now supported for rustls via https://github.com/kube-rs/kube/pull/1104 thanks to @MikailBag 's initial PR and it's released in 0.77.0.

Unfortunately, it is not done for openssl yet (where the upstream pr in hyper-openssl is not responded to) and I couldn't find a way to do it directly .

**Comment by goenning** [comment]
That crate doesn't seem to be very active, is forking hyper-openssl an option until that PR is merged? 🤔

**Comment by clux** [comment]
Someone would have to step up to do that type of slog though.

As a potential alternative; we just closed all the rustls issues on `main` after the last rustls/hyper-rustls bump. So switching to `rustls-tls` feature should be more viable for apps now.

**Comment by goenning** [comment]
I'll give it a go, I recall seeing some caveats on README which made me pick openssl instead.

**Comment by yann-soubeyrand** [comment]
Hello, is there some news on this front?

**Comment by clux** [comment]
Context on bug since this is old. This is supported on `rustls` for us which is our default stack since [0.86.0](https://github.com/kube-rs/kube/releases/tag/0.86.0) / sep 2023.

The openssl support is missing and help is **still welcome** on this front from anyone actually using this stack.

**Comment by dgunzy** [comment]
Hey @clux  - I opened #1993  which should close it - i tested this as a patch to help fix my issue in  [flux9s#170](https://github.com/dgunzy/flux9s/issues/170) and it worked well.  First time contributing, so open to PR feedback on what could be improved there.
